crates/autopilot/src/on_settlement_event_updater.rs

```diff
@@ -59,30 +67,52 @@ enum AuctionIdRecoveryStatus {
 }
 
 impl OnSettlementEventUpdater {
-    pub async fn run_forever(self) -> ! {
-        let mut current_block = self.eth.current_block().borrow().to_owned();
-        let mut block_stream = ethrpc::current_block::into_stream(self.eth.current_block().clone());
+    pub fn new(eth: infra::Ethereum, db: Postgres) -> Self {
+        let inner = Arc::new(Inner {
+            eth,
+            db,
+            notify: Notify::new(),
+        });
+        let inner_clone = inner.clone();
+        tokio::spawn(async move { Inner::wait_for_updates(inner_clone).await });
+        Self { inner }
+    }
+
+    /// Deletes settlement_observations and order executions for the given range
+    pub async fn delete_observations(
+        transaction: &mut PgTransaction<'_>,
+        range: RangeInclusive<u64>,
+    ) -> Result<()> {
+        database::settlements::delete(transaction, *range.start() as i64)
+            .await
+            .context("delete_settlement_observations")?;
+
+        Ok(())
+    }
+
+    /// Schedules an update loop on a background thread
+    pub fn schedule_update(&self) {
+        self.inner.notify.notify_one();
+    }
+}
+
+impl Inner {
+    async fn wait_for_updates(self: Arc<Inner>) -> ! {
         loop {
             match self.update().await {
                 Ok(true) => {
-                    tracing::debug!(
-                        block = current_block.number,
-                        "on settlement event updater ran and processed event"
-                    );
-                    // Don't wait until next block in case there are more pending events to process.
+                    tracing::debug!("on settlement event updater ran and processed event");
+                    // There might be more pending updates, continue immediately.
                     continue;
                 }
                 Ok(false) => {
-                    tracing::debug!(
-                        block = current_block.number,
-                        "on settlement event updater ran without update"
-                    );
+                    tracing::debug!("on settlement event updater ran without update");
                 }
                 Err(err) => {
                     tracing::error!(?err, "on settlement event update task failed");
                 }
             }
-            current_block = block_stream.next().await.expect("blockchains never end");
+            self.notify.notified().await;
```

Comment:
I think the we should only have to wait for another notification if we are currently not aware of additional work (i.e. in the `Ok(false)`) case. Otherwise when we have to index a bunch of work at once (possibly after a restart) we'll never catch up. This would also mean that we can get rid of the `continue`s.

Reply:
There is a `continue` in the `Ok(true)` branch making us skip the wait in this case. Would you prefer to remove the continue and instead move the `notify` into the two unsuccessful branches?

Reply:
Ah, never mind. Must have overlooked that.